VestingSchedule.sol

```diff
@@ -0,0 +1,227 @@
+        _beneficiary = beneficiary;
+        _vestingAmount = vestingAmount;
+        _revocable = revocable;
+        _duration = vestingPeriodSec;
+        _cliff = vestingStartTime.add(vestingCliff);
+        _start = vestingStartTime;
+        _refundDestination = refundDestination;
+        _revoker = revoker;
+    }
+
+    /**
+     * @return the beneficiary of the tokens.
+     */
+    function beneficiary() external view returns (address) {
+        return _beneficiary;
+    }
+
+    /**
+     * @return the amount to be vested.
+     */
+    function vestingAmount() external view returns (uint256) {
+        return _vestingAmount;
+    }
+
+    /**
+     * @return the cliff time of the token vesting.
+     */
+    function cliff() external view returns (uint256) {
+        return _cliff;
+    }
+
+    /**
+     * @return the start time of the token vesting.
+     */
+    function start() external view returns (uint256) {
+        return _start;
+    }
+
+    /**
+     * @return the duration of the token vesting.
+     */
+    function duration() external view returns (uint256) {
+        return _duration;
+    }
+
+    /**
+     * @return true if the vesting is revocable.
+     */
+    function revocable() external view returns (bool) {
+        return _revocable;
+    }
+
+    /**
+     * @return the cummulative amount of tokens released.
+     */
+    function released() external view returns (uint256) {
+        return _released;
+    }
+
+    /**
+     * @return true if the vesting is revoked.
+     */
+    function revoked(address token) external view returns (bool) {
+        return _revoked;
+    }
+
+    /**
+     * @return the timestamp of the revocation.
+     */
+    function revokeTime(address token) external view returns (bool) {
+        return _revokeTime;
+    }
+
+    /**
+     * @notice Transfers available released tokens from the vesting back to beneficiary.
+     */
+    function withdraw() external onlyBeneficiary {
+        uint256 unreleased = _releasableAmount(block.timestamp);
+
+        require(unreleased > 0, "No unreleased tokens are due for withdraw");
+
+        _released = _released.add(unreleased);
+
+        token.safeTransfer(_beneficiary, unreleased);
+
+        emit VestingWithdrawn(msg.sender, unreleased, block.timestamp);
+    }
+
+    /**
+     * @notice Allows only the revoker to revoke the vesting. Tokens already vested
+     * remain in the contract, the rest are returned to the _refundDestination.
+     * @param revokeTime the revocation timestamp
+     * @dev revokeTime the revocation timestamp. If is less than the current block timestamp, it is set equal
+     */
+    function revoke(revokeTime) external onlyRevoker {
+        require(_revocable, "Revoking is not allowed");
+        require(!_revoked, "Vesting already revoked");
+
+        uint256 revokeTimestamp = revokeTime > block.timestamp ? revokeTime : block.timestamp;
+
+        uint256 balance = getGoldToken().balanceOf(address(this));
```

Comment:
I am not super sure we should be basing our logic on the current token balance. I'd think it should be somehow based on the totalVestedAmount

Reply:
implemented in the logic of ht enew revision